README.md
<p align="center">
  <a href="https://www.npmjs.org/package/trickling">
    <img src="https://img.shields.io/npm/v/trickling.svg">
  </a>
  <a href="https://npmcharts.com/compare/trickling?minimal=true">
    <img src="https://img.shields.io/npm/dm/trickling.svg">
  </a>
  <br>
</p>

# Trickling

A modern progress bar for web APP. Featuring realistic trickle animations to convince your users that something is happening!

- [See Example here](https://hongwenqing.com/trickling/).
- [Release Notes](./CHANGELOG.md).

# Features

- Support **programmatically** modify the style of the progress bar.
- Support **RTL** languages.
- Support **Typescript**.

# Installation

```bash
# pnpm
$ pnpm add trickling

# yarn
$ yarn add trickling

# npm
$ npm i trickling
```

# Usage

1. Imports style.

```ts
import 'trickling/lib/style.css'
```

2. Create a trickling progress instance in a single file.

```ts
// @/plugins/trickling-progress.ts

import { createTrickling } from 'trickling'

// Create a Trickling progress instance
const tricklingProgress = createTrickling({
  // Options
  // ...
})

// You can also change options after creating a Trickling progress instance
tricklingProgress.setOptions({
  // ...
})

// Export the instance
export { tricklingProgress }
```

3. Using it where needed.

```ts
import { tricklingProgress } from '@/plugins/trickling-progress.ts'

// ...
// shows the Trickling progress bar
tricklingProgress.start()

// ...
// Then, completes the Trickling progress
tricklingProgress.done()
```

# Options

| Key | Type | Default value | Description |
| :---: | :---: | :---: | :---: |
| `minimum` | `number` | `0.08` | Changes the minimum percentage used upon starting. |
| `maximum` | `number` | `0.994` | Changes the maximum percentage used upon ending. |
| `easing` | `string` | `ease` | Adjust animation settings using easing (a CSS easing string). |
| `speed` | `number` | `200` | Adjust animation settings using speed (in ms). |
| `trickle` | `boolean` | `true` | Turn off the automatic incrementing behavior by setting this to `false`. |
| `trickleSpeed` | `number` | `1000` | Adjust how often to trickle/increment (in ms). |
| `showSpinner` | `boolean` | `true` | Turn off **loading spinner** by setting it to `false`. |
| `appendTo` | `string`, `HTMLElement` | `body` | Specify this to change the **parent container**. |
| `customWrapperClassName` | `string` | `''` | Specify this to add a class name into the parent container. |
| `color` | `string` | `#2299dd` | Specify this to change **color** of the progress bar and spinner. |
| `progressBarHeight` | `string` | `2px` | Specify this to change **height** of the progress bar. |
| `spinnerOpacity` | `number` | `1` | Specify this to change **opacity** of the loading spinner. |
| `spinnerSize` | `string` | `18px` | Specify this to change **size** of the loading spinner. |
| `spinnerStrokeWidth` | `string` | `2px` | Specify this to change **stroke width** of the loading spinner. |
| `rtl` <br /> (Added in v1.6.0) | `boolean` | `false` | Change the progress direction to right-to-left. |
| `removeFromDOMWhenDone` <br /> (Added in v1.9.0) | `boolean` | `true` | Remove the component from the DOM when done, re-add when needed. This can have performance implications on complex apps as style calculations are slow. If set to `false`, just hidden the DOM via `display: none` when progress done. |
| `zIndex` <br /> (Added in v1.8.0) | `number/string` | `1031` | Specify this to change progress bar **z-index**. |
| [trickleIncrementalCurve](https://github.com/yisibell/trickling#trickle-incremental-curve) <br /> (Added in v1.10.0) | `function/Array` | See below | You can use this option to configure the **incremental curve** of the trickle. |


# Trickle Incremental Curve

Using this configuration allows you to easily change the default trickle increasing curve. Thus, each Progress bar has different incremental expressions.

1. Default value

```ts
[
  { from: 0, to: 0.2, value: 0.1 },
  { from: 0.2, to: 0.5, value: 0.04 },
  { from: 0.5, to: 0.8, value: 0.02 },
  { from: 0.8, to: 0.99, value: 0.005 },
]
```

- **from**: This represents the starting range of the current progress status (including).
- **to**: This represents the end range of the current progress status (excluding).
- **value**: This represents the progress increment value of the current range.

2. Type definition see [here](./src/lib/interfaces/core.ts).

- **Function**: If you use it as a function, should returns a `number` or `array`.
- **Array**: Just set it via a `array`.

# Advanced usage

1. **Percentages**: To set a progress percentage, call `.set(n)`, where n is a number between `0..1`.

```ts
trickling.set(0.0);     // Just same as .start()
trickling.set(0.4);
trickling.set(1.0);     // Just same as .done()
```

2. **Incrementing**: To increment the progress bar, just use `.inc()`. This increments it with a random amount. This will never get to `100%`: use it for every image load (or similar).

```ts
trickling.inc();
```

3. If you want to increment by a specific value, you can pass that as a parameter:

```ts
trickling.inc(0.2);    // This will get the current status value and adds 0.2 until status is 0.994
```

4. **Force-done**: By passing true to `done()`, it will show the progress bar even if it's not being shown. (The default behavior is that `.done()` will not do anything if `.start()` isn't called).

```ts
trickling.done(true);
```

5. Get the status value: To get the status value, use `.status`.

```ts
trickling.status
```

CHANGELOG.md
# Changelog


## v1.14.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.13.1...v1.14.0)

### 🚀 Enhancements

- Upgrade all dev deps version to latest ([602e7f3](https://github.com/yisibell/trickling/commit/602e7f3))

### 🏡 Chore

- **example:** Update prismjs highlight init ([034994b](https://github.com/yisibell/trickling/commit/034994b))
- Update readme ([0d8ed26](https://github.com/yisibell/trickling/commit/0d8ed26))

### ❤️ Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.13.1

[compare changes](https://github.com/yisibell/trickling/compare/v1.13.0...v1.13.1)

### 🩹 Fixes

- TrickleIncrementalCurve types ([4591579](https://github.com/yisibell/trickling/commit/4591579))

### ❤️ Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.13.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.12.0...v1.13.0)

### 🚀 Enhancements

- Upgrade typescript to v5 and other non-major deps ([b8f6f49](https://github.com/yisibell/trickling/commit/b8f6f49))

### ❤️ Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.12.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.11.0...v1.12.0)


### 🚀 Enhancements

  - Extract `dom` into the `fourdom` package ([950bb93](https://github.com/yisibell/trickling/commit/950bb93))

### 🏡 Chore

  - Sync lib dir ([27acbc2](https://github.com/yisibell/trickling/commit/27acbc2))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.11.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.10.0...v1.11.0)


### 🚀 Enhancements

  - TrickleIncrementalCurve function support returns a number ([a55a80e](https://github.com/yisibell/trickling/commit/a55a80e))

### 🏡 Chore

  - **docs:** Update readme ([9554527](https://github.com/yisibell/trickling/commit/9554527))
  - **docs:** Update readme ([2570105](https://github.com/yisibell/trickling/commit/2570105))
  - **docs:** Update readme ([599266e](https://github.com/yisibell/trickling/commit/599266e))
  - **docs:** Update readme ([352b5b5](https://github.com/yisibell/trickling/commit/352b5b5))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.10.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.9.0...v1.10.0)


### 🚀 Enhancements

  - Support trickleIncrementalCurve option to configure the incremental curve of the trickle ([9e65890](https://github.com/yisibell/trickling/commit/9e65890))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.9.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.8.3...v1.9.0)


### 🚀 Enhancements

  - Change removeFromDOM to removeFromDOMWhenDone ([27444d1](https://github.com/yisibell/trickling/commit/27444d1))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.8.3

[compare changes](https://github.com/yisibell/trickling/compare/v1.8.2...v1.8.3)


### 🏡 Chore

  - **docs:** Update readme and example ([3e19926](https://github.com/yisibell/trickling/commit/3e19926))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.8.2

[compare changes](https://github.com/yisibell/trickling/compare/v1.8.1...v1.8.2)


### 🏡 Chore

  - Sync lib dir ([6590632](https://github.com/yisibell/trickling/commit/6590632))
  - **example:** Update example style ([372c307](https://github.com/yisibell/trickling/commit/372c307))
  - **docs:** Update docs and example ([4be43cb](https://github.com/yisibell/trickling/commit/4be43cb))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.8.1

[compare changes](https://github.com/yisibell/trickling/compare/v1.8.0...v1.8.1)


### 🩹 Fixes

  - Reset progress bar status to 0 when progress done in removeFromDOM config, to fix transition ([91f6f98](https://github.com/yisibell/trickling/commit/91f6f98))

### 🏡 Chore

  - Sync lib dir ([b352b0e](https://github.com/yisibell/trickling/commit/b352b0e))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.8.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.7.0...v1.8.0)


### 🚀 Enhancements

  - Support define z-index of progress bar DOM ([fbdf175](https://github.com/yisibell/trickling/commit/fbdf175))

### 🏡 Chore

  - **docs:** Update readme ([6dba53e](https://github.com/yisibell/trickling/commit/6dba53e))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.7.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.6.0...v1.7.0)


### 🚀 Enhancements

  - Support settings removeFormDOM to improve performance on IE 11 ([a3cfc0f](https://github.com/yisibell/trickling/commit/a3cfc0f))

### 🏡 Chore

  - Sync lib to repo ([fa24643](https://github.com/yisibell/trickling/commit/fa24643))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.6.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.5.0...v1.6.0)


### 🚀 Enhancements

  - Support right-to-left progress(for RTL languages) ([cfbabf4](https://github.com/yisibell/trickling/commit/cfbabf4))
  - More specific class name to prevent clashes ([e4e4bdb](https://github.com/yisibell/trickling/commit/e4e4bdb))

### 🏡 Chore

  - **docs:** Update readme ([6a18ff2](https://github.com/yisibell/trickling/commit/6a18ff2))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.5.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.4.1...v1.5.0)


### 🚀 Enhancements

  - Update build ([a175d1a](https://github.com/yisibell/trickling/commit/a175d1a))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.4.1

[compare changes](https://github.com/yisibell/trickling/compare/v1.4.0...v1.4.1)


### 🩹 Fixes

  - Style.css exports ([b812f70](https://github.com/yisibell/trickling/commit/b812f70))

### 🏡 Chore

  - Remove example favicon ([2c9d418](https://github.com/yisibell/trickling/commit/2c9d418))

### ❤️  Contributors

- Wenqing <wenqing@kerrylan.com>

## v1.4.0

[compare changes](https://github.com/yisibell/trickling/compare/v1.3.0...v1.4.0)


### 🚀 Enhancements


lib/trickling.js
var P = Object.defineProperty;
var k = (r, e, t) => e in r ? P(r, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : r[e] = t;
var p = (r, e, t) => (k(r, typeof e != "symbol" ? e + "" : e, t), t);
import { css as l, addClass as c, removeElement as u, removeClass as d } from "fourdom";
function f(r, e, t) {
  return r < e ? e : r > t ? t : r;
}
function y(r, e) {
  return (e ? (1 - r) * 100 : (-1 + r) * 100).toFixed(4);
}
const E = () => {
  const r = [];
  function e() {
    const t = r.shift();
    t && t(e);
  }
  return function(t) {
    r.push(t), r.length == 1 && e();
  };
}, b = E();
var o = /* @__PURE__ */ ((r) => (r.wrapperSelectorId = "trickling", r.customParentClassName = "trickling-custom-parent", r.busyFlagClassName = "trickling-busy", r.template = `
  <div class="trickling-progress-bar" role="bar">
    <div class="trickling-progress-peg"></div>
  </div>
  <div class="trickling-progress-spinner" role="spinner">
    <div class="trickling-progress-spinner__spinner-icon"></div>
  </div>`, r.barSelector = '[role="bar"]', r.spinnerSelector = '[role="spinner"]', r.rtlClassName = "rtl", r))(o || {}), h = /* @__PURE__ */ ((r) => (r.queryBarElementError = "[Trickling]: Can not find 'barSelector' element!", r.queryAppendToElementError = "[Trickling]: Can not find 'options.appendTo' element!", r))(h || {});
const a = class a {
  constructor(e) {
    p(this, "progressOffsetWidth", 0);
    p(this, "status", null);
    p(this, "positionUsing", "");
    p(this, "options", {
      speed: 200,
      easing: "ease",
      appendTo: "body",
      customWrapperClassName: "",
      minimum: 0.08,
      maximum: 0.994,
      showSpinner: !0,
      trickleSpeed: 1e3,
      trickle: !0,
      color: "#2299dd",
      progressBarHeight: "2px",
      spinnerOpacity: 1,
      spinnerSize: "18px",
      spinnerStrokeWidth: "2px",
      zIndex: 1031,
      rtl: !1,
      removeFromDOMWhenDone: !0,
      trickleIncrementalCurve: [
        { from: 0, to: 0.2, value: 0.1 },
        { from: 0.2, to: 0.5, value: 0.04 },
        { from: 0.5, to: 0.8, value: 0.02 },
        { from: 0.8, to: 0.99, value: 5e-3 }
      ]
    });
    this.options = Object.assign(this.options, e), this.setPercent(null);
  }
  setOptions(e) {
    return this.options = Object.assign(this.options, e), this;
  }
  setCSSVars(e) {
    l(e, {
      "--trickling-color": this.options.color,
      "--trickling-progress-bar-height": this.options.progressBarHeight,
      "--trickling-spinner-opacity": this.options.spinnerOpacity,
      "--trickling-spinner-size": this.options.spinnerSize,
      "--trickling-spinner-stroke-width": this.options.spinnerStrokeWidth,
      "--trickling-progress-bar-z-index": this.options.zIndex
    });
  }
  render(e) {
    if (this.isRendered())
      return this.getWrapperElement();
    c(document.documentElement, o.busyFlagClassName);
    const t = document.createElement("div");
    t.id = o.wrapperSelectorId, t.innerHTML = o.template, this.options.rtl && c(t, o.rtlClassName), this.options.customWrapperClassName && c(t, this.options.customWrapperClassName), this.setCSSVars(t);
    const s = this.getBarElement(t), n = this.getAppendToElement();
    if (this.translateProgressBar(s, e), !this.options.showSpinner) {
      const i = t.querySelector(o.spinnerSelector);
      i && u(i);
    }
    return n != document.body && c(n, o.customParentClassName), n && n.appendChild(t), t;
  }
  triggerRepaint() {
    const e = this.getWrapperElement();
    e && (this.progressOffsetWidth = e.offsetWidth);
  }
  set(e) {
    const t = this.isStarted();
    !t && this.visible(), e = f(e, this.options.minimum, 1), this.setPercent(e === 1 ? null : e);
    const s = this.render(!t), n = this.getBarElement(s), i = this.options.speed, v = this.options.easing;
    return this.triggerRepaint(), b((g) => {
      this.positionUsing === "" && (this.positionUsing = this.getPositioningCSS()), l(n, this.barPositionCSS(e, i, v)), e === 1 ? (l(s, {
        transition: "none",
        opacity: 1
      }), this.triggerRepaint(), setTimeout(() => {
        l(s, {
          transition: `all ${i}ms linear`,
          opacity: 0
        }), setTimeout(() => {
          this.remove(), g();
        }, i);
      }, i)) : setTimeout(g, i);
    }), this;
  }
  inc(e) {
    let t = this.getPercent();
    if (t) {
      if (t > 1)
        return this;
      if (typeof e != "number") {
        const s = typeof this.options.trickleIncrementalCurve == "function" ? this.options.trickleIncrementalCurve(t) : this.options.trickleIncrementalCurve;
        if (typeof s == "number")
          e = s || 0;
        else {
          const n = s.find(
            (i) => t !== null && t >= i.from && t < i.to
          );
          n ? e = n.value : e = 0;
        }
      }
      return t = f(t + e, 0, this.options.maximum), this.set(t);
    } else
      return this.start();
  }
  trickle() {
    return this.inc();
  }
  start() {
    !this.isStarted() && this.visible(), this.getPercent() || this.set(0);
    const e = () => {
      setTimeout(() => {
        this.getPercent() && (this.trickle(), e());
      }, this.options.trickleSpeed);
    };
    return this.options.trickle && e(), this;
  }
  done(e) {
    return !e && !this.getPercent() ? this : this.inc(0.3 + 0.5 * Math.random()).set(1);
  }
  translateProgressBar(e, t) {
    const s = t ? this.getBarPercentage(0) : this.getBarPercentage(this.getPercent() || 0);
    l(e, {
      transition: "all 0 linear",
      transform: `translate3d(${s}%, 0, 0)`
    });
  }
  visible() {
    if (this.isRendered() && !this.options.removeFromDOMWhenDone) {
      const e = this.getWrapperElement();
      e && l(e, { display: "block", opacity: 1 }), this.setPercent(null);
    }
  }
  hidden() {
    const e = this.getWrapperElement();
    e && (l(e, { display: "none" }), this.setPercent(null), this.translateProgressBar(this.getBarElement(e), !0));
  }
  remove(e) {
    if (d(document.documentElement, o.busyFlagClassName), !this.options.removeFromDOMWhenDone && !e) {
      this.hidden();
      return;
    }
    const t = this.getAppendToElement();
    d(t, o.customParentClassName);
    const s = this.getWrapperElement();
    s && u(s);
  }
  getWrapperElement() {
    return document.getElementById(o.wrapperSelectorId);
  }
  getBarElement(e) {
    const t = e.querySelector(o.barSelector);
    if (!t)
      throw new Error(h.queryBarElementError);
    return t;
  }
  getAppendToElement() {
    const e = typeof this.options.appendTo == "string" ? document.querySelector(this.options.appendTo) : this.options.appendTo;
    if (!e)
      throw new Error(h.queryAppendToElementError);
    return e;
  }
  setPercent(e) {
    this.status = e;
  }
  getPercent() {
    return this.status;
  }
  isRendered() {
    return !!this.getWrapperElement();
  }
  isStarted() {
    return typeof this.getPercent() == "number";
  }
  getBarPercentage(e) {
    return y(e, this.options.rtl);
  }
  barPositionCSS(e, t, s) {
    let n = {};
    return this.positionUsing === "translate3d" ? n = {
      transform: `translate3d(${this.getBarPercentage(e)}%,0,0)`
    } : this.positionUsing === "translate" ? n = {
      transform: `translate(${this.getBarPercentage(e)}%,0)`
    } : n = { "margin-left": `${this.getBarPercentage(e)}%` }, n.transition = `all ${t}ms ${s} 0s`, n;
  }
  getPositioningCSS() {
    const e = document.body.style, t = "WebkitTransform" in e ? "Webkit" : "MozTransform" in e ? "Moz" : "msTransform" in e ? "ms" : "OTransform" in e ? "O" : "";
    return t + "Perspective" in e ? "translate3d" : t + "Transform" in e ? "translate" : "margin";
  }
  static createProgress(e) {
    return a.instance || (a.instance = new a(e)), a.instance;
  }
};
p(a, "instance");
let m = a;
const B = function(r) {
  return m.createProgress(r);
};
export {
  m as Trickling,
  B as createTrickling
};
